athena-gcs/src/main/java/com/amazonaws/athena/connectors/gcs/GcsMetadataHandler.java

```diff
@@ -0,0 +1,320 @@
+    // used to filter out Glue tables which lack indications of being used for DDB.
+    private static final TableFilter TABLE_FILTER = (Table table) -> table.getStorageDescriptor().getLocation().contains(TABLE_FILTER_IDENTIFIER)
+            || (table.getParameters() != null && GcsUtil.isSupportedFileType(table.getParameters().get("classification")))
+            || (table.getStorageDescriptor().getParameters() != null && GcsUtil.isSupportedFileType(table.getStorageDescriptor().getParameters().get("classification")));
+    private final StorageMetadata datasource;
+    private final AWSGlue glueClient;
+
+    public GcsMetadataHandler() throws IOException
+    {
+        super((System.getenv(GLUE_ENV) != null && !"false".equalsIgnoreCase(System.getenv(GLUE_ENV))), SOURCE_TYPE);
+        String gcsCredentialsJsonString = getGcsCredentialJsonString(this.getSecret(System.getenv(GCS_SECRET_KEY_ENV_VAR)), GCS_CREDENTIAL_KEYS_ENV_VAR);
+        this.datasource = createDatasource(gcsCredentialsJsonString, System.getenv());
+        this.glueClient = getAwsGlue();
+    }
+
+    @VisibleForTesting
+    @SuppressWarnings("unused")
+    protected GcsMetadataHandler(EncryptionKeyFactory keyFactory,
+                                 AWSSecretsManager awsSecretsManager,
+                                 AmazonAthena athena,
+                                 String spillBucket,
+                                 String spillPrefix,
+                                 AmazonS3 amazonS3, AWSGlue glueClient) throws IOException
+    {
+        super(glueClient, keyFactory, awsSecretsManager, athena, SOURCE_TYPE, spillBucket, spillPrefix);
+        String gcsCredentialsJsonString = getGcsCredentialJsonString(this.getSecret(System.getenv(GCS_SECRET_KEY_ENV_VAR)), GCS_CREDENTIAL_KEYS_ENV_VAR);
+        this.datasource = createDatasource(gcsCredentialsJsonString, System.getenv());
+        this.glueClient = getAwsGlue();
+        System.setProperty(SDKGlobalConfiguration.DISABLE_CERT_CHECKING_SYSTEM_PROPERTY, "true");
+    }
+
+    /**
+     * Used to get the list of schemas (aka databases) that this source contains.
+     *
+     * @param allocator Tool for creating and managing Apache Arrow Blocks.
+     * @param request   Provides details on who made the request and which Athena catalog they are querying.
+     * @return A ListSchemasResponse which primarily contains a Set<String> of schema names and a catalog name
+     * corresponding the Athena catalog that was queried.
+     */
+    @Override
+    public ListSchemasResponse doListSchemaNames(BlockAllocator allocator, ListSchemasRequest request) throws Exception
+    {
+        Set<String> schema = new LinkedHashSet<>();
+        if (glueClient != null) {
+            try {
+                schema.addAll(super.doListSchemaNames(allocator, request, DB_FILTER).getSchemas());
```

Comment:
why do we add extra layer or Set<String> schema here? can't we just return directly from super.doListSchemaNames method call? Or do we intended to remove the duplicate table name using Set?

Reply:
refactored with the required changes.